client/login/wp-login/index.js

```diff
@@ -2,34 +2,139 @@
-		</div>
-		<div className="wp-login__container">
-			<LoginBlock
-				title={ translate( 'Sign in to WordPress.com' ) } />
-		</div>
-		<div className="wp-login__footer">
-			{
-				isEnabled( 'magic-login' ) && (
-					<a href="/login/send-me-a-link">{ translate( 'Email me a login link' ) }</a>
-				)
-			}
-		</div>
-	</Main>
-);
-
-export default localize( Login );
+class Login extends React.Component {
+	magicLoginMainContent() {
+		const {
+			handlingMagicLink,
+			magicLoginEmailAddress,
+			showingCheckYourEmailPage,
+			showingMagicLoginExpiredPage,
+		} = this.props;
+
+		if ( showingMagicLoginExpiredPage ) {
+			return <EmailedLoginLinkExpired />;
+		}
+
+		if ( showingCheckYourEmailPage ) {
+			return <EmailedLoginLinkSuccessfully emailAddress={ magicLoginEmailAddress } />;
+		}
+
+		if ( handlingMagicLink ) {
+			return <HandleEmailedLinkForm />;
+		}
+	}
+
+	footerContent() {
+		const {
+			handlingMagicLink,
+			showingCheckYourEmailPage,
+			showingMagicLoginRequestForm,
+			translate,
+		} = this.props;
+
+		if ( ! (
+			handlingMagicLink ||
+			showingCheckYourEmailPage ||
+			showingMagicLoginRequestForm
+		) ) {
+			return <a href="#" onClick={ this.props.onMagicLoginRequestClick }>{ translate( 'Email me a login link' ) }</a>;
+		}
+	}
+
+	render() {
+		const {
+			showingMagicLoginRequestForm,
+			translate,
+		} = this.props;
+
+		return (
+			<Main className="wp-login">
+				{ this.magicLoginMainContent() || (
+					<div>
+						<div className="wp-login__header">
+							<Gridicon icon="user-circle" size={ 72 } />
+							<div>{
+								// @TODO show currently logged in user if any
+								translate( 'You are signed out' )
+							}</div>
+						</div>
+						<div className="wp-login__container">
+							{ showingMagicLoginRequestForm
+								? <RequestLoginEmailForm />
+								: <LoginBlock title={ translate( 'Sign in to WordPress.com' ) } />
+							}
+						</div>
+						<div className="wp-login__footer">
+							{ this.footerContent() }
+						</div>
+					</div>
+				) }
+			</Main>
+		);
+	}
+}
+
+const mapState = state => {
+	const magicLoginEnabled = isEnabled( 'magic-login' );
+	const queryArguments = getCurrentQueryArguments( state );
+
+	const {
+		action,
+		client_id: clientId,
+		email,
+		token,
+		tt: tokenTime,
+	} = queryArguments;
+
+	return {
+		handlingMagicLink: (
+			magicLoginEnabled &&
+			action === 'handleLoginEmail' &&
+			clientId &&
+			email &&
+			token &&
+			tokenTime
+		),
+		magicLoginEmailAddress: emailAddressFormInput( state ),
+		showingMagicLoginExpiredPage: magicLoginEnabled && isShowingMagicLoginExpiredPage( state ),
+		showingMagicLoginRequestForm: magicLoginEnabled && isShowingMagicLoginRequestForm( state ),
+		showingCheckYourEmailPage: magicLoginEnabled && isShowingCheckYourEmailPage( state ),
+	};
+};
+
+const mapDispatch = dispatch => {
+	return {
+		onMagicLoginRequestClick: event => {
+			event.preventDefault();
+			dispatch( showMagicLoginRequestForm() );
+		}
```

Comment:
You should only use `mapDispatch` to wrangle bindings of state, props and dispatch. The presence of `event` here is a code smell to me: this a function that should be defined in the component. So:

```js
const mapDispatch = {
  showMagicLoginRequestForm
};
```

Then in the component class:

```js
onMagicLoginRequestClick = ( event ) => {
  event.preventDefault();
  this.props.showMagicLoginRequestForm();
}
```